### ISSUE TITLE
RFC: Huge Refactor + Feature Additions

### ISSUE THREAD

**Issue by nateabele** [angular-ui/ui-router#286]
_Brace yourselves. This is pretty freakin' long._
## I Have a Branch to Which I Have Committed A Whole Bunch of Crap

For reference, [see here](https://github.com/nateabele/ui-router/tree/viewpath). We'll be talking about what's going on in those last few commits (the ones with my name on them). Let's just go straight down the list, shall we?
## State-Building Refactor

The `registerState()` method, which is used to... register new states, was previously pretty large. I have refactored it into a [state builder object](https://github.com/nateabele/ui-router/blob/viewpath/src/state.js#L7-L109), which makes the various aspects of constructing a state much more comprehensible and easy to extend (albeit only internally for now... maybe we can leverage this in the provider API to allow people to generate or auto-assign their own state attributes... I literally thought of that just now). It also makes `registerState()` around 20 lines of code, which is nice.
## The ^ Character

We have a new magical character: `^`. This character is basically used to append or inherit from the current state when transitioning. You can check out the [tests which demonstrate this functionality](https://github.com/nateabele/ui-router/blob/viewpath/test/stateSpec.js#L213-L236) using the new `$state.go()` short-hand method. You can transition to a `^child` state, a `^.sibling` state, or a `^.^.parent` state. This syntax is implemented directly in `findState()`, so it should work for anything that accepts a state name, however, any calls to `findState()` must be refactored to pass the state that the lookup is relative to. I also plan to make the `^` character for inheriting parent URLs, i.e. `url:`^/child`, because the current functionality of appending to the parent URL automatically is confusing and inflexible.

Also, you may notice that those tests demonstrate...
## Cascading Parameter Inheritance

Yup, that's a thing now. It happens automatically when using `$state.go()`, or when passing `{ inherit: true }` to `$state.transitionTo()` (and yup, that's been refactored to accept an options hash as its third parameter).
## The `$view` service

The template-management code has been refactored into the new `$view` service. This code is still pretty crap, and needs some work, but at least it's not directly embedded inside `$state` anymore. I also still need to write some tests for it, so there's that, too.
## View Targeting

I'm kind of lying by including this in the list, because it's actually not done (yet), but will be committed soon. In short, the current approach to view targeting is crap. It is the single most confusing aspect of what is an otherwise fairly straightforward library. We've gotten more questions on that, than on any other single aspect of this library (except maybe for resolves, but those fixes are already in progress).

Here's the thing: if I'm targeting a view, I should be able to use a simple dot-syntax, i.e. `viewA.viewB.viewC`, regardless of what state created that view. If a named view exists, then it exists. If it doesn't, it doesn't. Which state created it is totally irrelevant. This also makes view targeting much more flexible, as we can do things like allow relative targeting with `^`, similar to the above. This is going to get way more important as people try to use this library to scale out projects, or create redistributable 3rd-party parts of applications, which may be "mounted" at arbitrary parent states within an application.
## In Summary

I didn't want to submit a PR for this until we had a chance to discuss it, and because it's still in progress, but I wanted to collect everyone's thoughts on this as I wrap up these changes. Thanks for reading. /cc @ksperling, @timkindberg, @ProLoser, @stu-salsbury, @eahutchins
